File: python/house_robber.py
class Solution:
    def rob(self, nums: list[int]) -> int:
        """time complexity: O(n)"""
        if len(nums) == 0:
            return 0
        if len(nums) == 1:
            return nums[0]
        dp: list[int] = [0] * len(nums)
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])
        for i in range(2, len(nums)):
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[i])
        return dp[-1]


def test_rob_case_1():
    # arragne
    nums: list[int] = [1, 2, 3, 1]
    expected: int = 4

    # act
    solution = Solution()
    actual = solution.rob(nums)

    # assert
    assert expected, actual


def test_rob_case_2():
    # arragne
    nums: list[int] = [2, 7, 9, 3, 1]
    expected: int = 12

    # act
    solution = Solution()
    actual = solution.rob(nums)

    # assert
    assert expected, actual

File: rust/src/code/house_robber.rs
pub struct Solution {}

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let mut rob1: i32 = 0;
        let mut rob2: i32 = 0;

        for n in nums {
            let new_rob = std::cmp::max(rob1 + n, rob2);
            rob1 = rob2;
            rob2 = new_rob;
        }

        rob2
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn rob_case_1() {
        // arrange
        let nums: Vec<i32> = vec![1, 2, 3, 1];
        let expected: i32 = 4;

        // act
        let actual = Solution::rob(nums);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn rob_case_2() {
        // arrange
        let nums: Vec<i32> = vec![2, 7, 9, 3, 1];
        let expected: i32 = 12;

        // act
        let actual = Solution::rob(nums);

        // assert
        assert_eq!(expected, actual);
    }
}
